<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>programs.phenotyper.AllelicDiversity &mdash; VLPB 0.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="VLPB 0.2 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li><a href="../../../index.html">VLPB 0.2 documentation</a> &raquo;</li>
          <li><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for programs.phenotyper.AllelicDiversity</h1><div class="highlight"><pre>
<span class="kn">from</span> <span class="nn">model.cluster</span> <span class="kn">import</span> <span class="n">Grid</span>
<span class="kn">from</span> <span class="nn">programs</span> <span class="kn">import</span> <span class="n">Haplotyper</span>
<span class="kn">import</span> <span class="nn">Readers</span><span class="o">,</span> <span class="nn">logging</span>


<div class="viewcode-block" id="AllelicDiversity"><a class="viewcode-back" href="../../../programs.phenotyper.html#programs.phenotyper.AllelicDiversity.AllelicDiversity">[docs]</a><span class="k">class</span> <span class="nc">AllelicDiversity</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;The class AllelicDiversity includes all methods and variables needed for calculation of the allelic diversity.</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pool</span><span class="p">,</span><span class="n">gffFile</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The constructor sets the given variables as instance variables</span>
<span class="sd">        </span>
<span class="sd">        :param pool: The pool to calculate the allelic diversity for</span>
<span class="sd">        :type pool: an instance of a :py:class`Pool.Pool` object</span>
<span class="sd">        :param gffFile: The path to the gff file</span>
<span class="sd">        :type gffFile: str. - path to the gffFile</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gffFile</span> <span class="o">=</span> <span class="n">gffFile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pool</span> <span class="o">=</span> <span class="n">pool</span>
    <span class="c">#&quot;C:/Users/Jetse/data/test/testPhenotyper/geneModelChr12.gff3&quot;, vcfFile = &quot;C:/Users/Jetse/data/test/testPhenotyper/fictiveSNPs.vcf&quot;</span>

    <span class="k">def</span> <span class="nf">_getAllHaplotypesByAccession</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contigs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The method getAllHaplotypesByAccession retrieves creates a dictionary with the accession as key and the haplotype as value</span>
<span class="sd">        </span>
<span class="sd">        :param contigs: The contigs to get the haplotypes from</span>
<span class="sd">        :type contigs: an list of :py:class:`Contig.Contig` instances</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">allHaplotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">:</span>
            <span class="n">haplotypes</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">snps</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">haplotype</span><span class="p">,</span><span class="n">accessions</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">accessions</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">haplotypes</span><span class="p">:</span>
                            <span class="n">haplotypes</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">haplotype</span><span class="p">)</span>   
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">haplotypes</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">haplotype</span><span class="p">]</span>
                <span class="n">allHaplotypes</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">haplotypes</span>
                
        <span class="k">return</span> <span class="n">allHaplotypes</span>
    
    <span class="k">def</span> <span class="nf">_parseFiles</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The method parseFiles creates the reader objects to parse all files.</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gffReader</span> <span class="o">=</span> <span class="n">Readers</span><span class="o">.</span><span class="n">GffReader</span><span class="p">()</span>
        <span class="n">gffReader</span><span class="o">.</span><span class="n">readFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gffFile</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span> <span class="o">=</span> <span class="n">gffReader</span><span class="o">.</span><span class="n">contigs</span>
        <span class="n">vcfReader</span> <span class="o">=</span> <span class="n">Readers</span><span class="o">.</span><span class="n">VcfReader</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="n">vcfReader</span><span class="o">.</span><span class="n">readFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vcfFile</span><span class="p">)</span>
        
<div class="viewcode-block" id="AllelicDiversity.getAllelicDiversity"><a class="viewcode-back" href="../../../programs.phenotyper.html#programs.phenotyper.AllelicDiversity.AllelicDiversity.getAllelicDiversity">[docs]</a>    <span class="k">def</span> <span class="nf">getAllelicDiversity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The method getAllelicDiversity calculates the allelic diversity and writes the output to a file.</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">Grid</span><span class="o">.</span><span class="n">useGrid</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">Haplotyper</span><span class="o">.</span><span class="n">executeBeagleCluster</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">Haplotyper</span><span class="o">.</span><span class="n">executeBeagleMultiThread</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">vcf</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">vcf</span><span class="p">:</span> 
            <span class="k">try</span><span class="p">:</span>
                <span class="n">outputFile</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">outputDir</span> <span class="o">+</span> <span class="s">&quot;/&quot;</span><span class="o">+</span><span class="n">vcf</span> <span class="o">+</span> <span class="s">&quot;_&quot;</span> <span class="o">+</span> <span class="s">&quot;allelicDiversity.csv&quot;</span>  
                <span class="bp">self</span><span class="o">.</span><span class="n">vcfFile</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">vcf</span><span class="p">[</span><span class="n">vcf</span><span class="p">]</span><span class="o">.</span><span class="n">getFile</span><span class="p">()</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_parseFiles</span><span class="p">()</span>
                <span class="n">haplotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getAllHaplotypesByAccession</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="p">)</span>
                <span class="n">accessions</span> <span class="o">=</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outWriter</span><span class="p">:</span>
                    <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;contig</span><span class="se">\t</span><span class="s">original</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">accessions</span><span class="p">:</span> <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span> <span class="n">accession</span> <span class="o">+</span> <span class="s">&quot;_1</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">+</span> <span class="n">accession</span> <span class="o">+</span> <span class="s">&quot;_2</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">contigId</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="p">:</span>
                        <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">contigId</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                        <span class="k">try</span><span class="p">:</span>
                            <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">allContigs</span><span class="p">[</span><span class="n">contigId</span><span class="p">]</span><span class="o">.</span><span class="n">refHaplotype</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span> <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;-</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">accessions</span><span class="p">:</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
                                <span class="k">if</span> <span class="n">contigId</span> <span class="ow">in</span> <span class="n">haplotypes</span><span class="p">:</span>
                                    <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[</span><span class="n">contigId</span><span class="p">][</span><span class="n">accession</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;-</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                        <span class="n">outWriter</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">ex</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">ex</span><span class="p">)</span></div></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li><a href="../../../index.html">VLPB 0.2 documentation</a> &raquo;</li>
          <li><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Jetse.
      Last updated on Oct 31, 2013.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>